src/Rings/mpoly-graded.jl

```diff
@@ -1612,6 +1612,349 @@ function Base.show(io::IO, h::HilbertData)
+  # The good ones will contribute to a minimal generating 
+  # set of the lhs ideal.
+  good = [e for e in shifted if all(x->(x>=0), e)]
+
+  # The bad monomials come from those which hop over the boundaries of 
+  # the monomial diagram by the shift. Their span has a new 
+  # generator which is collected in `bad` a priori. It is checked 
+  # whether they become superfluous and if not, they are added to 
+  # the good ones.
+  bad = [e for e in shifted if !all(x->(x>=0), e)]
+
+  divides(a::Vector{Int}, b::Vector{Int}) = all(k->(k>=0), b - a)
+
+  for e in bad 
+    m = [k < 0 ? 0 : k for k in e]
+    if all(x->(!divides(x, m)), good)
+      good = [x for x in good if !divides(m, x)]
+      push!(good, m)
+    end
+  end
+
+  return good
+end
+
+### This implements the speedup from the CoCoA strategy, 
+# see Exercise 4 in Section 5.3
+function _divide_by_monomial_power(a::Vector{Vector{Int}}, j::Int, k::Int)
+  divides(a::Vector{Int}, b::Vector{Int}) = all(k->(k>=0), b[1:j-1] - a[1:j-1]) && all(k->(k>=0), b[j+1:end] - a[j+1:end])
+  kept = [e for e in a if e[j] == 0]
+  for i in 1:k
+    next_slice = [e for e in a if e[j] == i]
+    still_kept = next_slice
+    for e in kept 
+      all(v->!divides(v, e), next_slice) && push!(still_kept, e)
+    end
+    kept = still_kept
+  end
+  still_kept = vcat(still_kept, [e for e in a if e[j] > k])
+
+  result = Vector{Vector{Int}}()
+  for e in still_kept
+    v = copy(e)
+    v[j] = (e[j] > k ? e[j] - k : 0)
+    push!(result, v)
+  end
+  return result
+end
+
+function _find_maximum(a::Vector{Int})
+  m = a[1]
+  j = 1
+  for i in 1:length(a)
+    if a[i] > m 
+      m = a[i]
+      j = i
+    end
+  end
+  return j
+end
+
+function _cmp_revlex(a::Vector{Int}, b::Vector{Int})
+  for i in 1:length(a)
+    a[i] == b[i] && continue
+    return a[i] > b[i]
+  end
+  return true
+end
+
+function _sort_lex(a::Vector{Vector{Int}})
+  pivot = a[1]
+  less = Vector{Vector{Int}}()
+  more = Vector{Vector{Int}}()
+  for m in a[2:end]
+    if _cmp_revlex(pivot, m)
+      push!(less, m)
+    else
+      push!(more, m)
+    end
+  end
+  return vcat(less, [pivot], more)
+end
+
+function _hilbert_numerator_from_leading_exponents(
+    a::Vector{Vector{Int}};
+    weight_matrix= [1 for i in 1:1, j in 1:length(a[1])],
+    return_ring=(all(x->(x>=0), weight_matrix) ? 
+                 PolynomialRing(QQ, [Symbol("t_$i") for i in 1:nrows(weight_matrix)])[1] : 
+                 LaurentPolynomialRing(QQ, [Symbol("t_$i") for i in 1:nrows(weight_matrix)])[1]
+                ),
+    #alg=:generator
+    #alg=:custom
+    #alg=:gcd
+    #alg=:indeterminate
+    #alg=:cocoa
+    alg=:BayerStillmanA # This is by far the fastest strategy. Should be used.
+  )
+  length(a) == 0 && return one(return_ring)
+
+  t = gens(return_ring)
+
+  # See Proposition 5.3.6
+  _are_pairwise_coprime(a) && return prod([1-prod([t[i]^(sum([e[j]*weight_matrix[i, j] for j in 1:length(e)])) for i in 1:length(t)]) for e in a])
```

Comment:
I would advise against writing such long convoluted lines. They are hard to read, understand and debug. As shown in other comments, often a simple loop is also faster; that may not be the case here, but I would recommend to at least change it to an `if ... end` and then apply some line wrapping

Reply:
This has been done at some internal point below. Since we're waiting for the MPolyBuildCtx, this is subject to change, anyway and can be copied from there when the time is ripe.